File: portscanner.js
const net = require('net');
const fs = require('fs');

function parsePorts(portArg) {
  if (portArg.includes('-')) {
    const [start, end] = portArg.split('-').map(Number);
    if (isNaN(start) || isNaN(end) || start > end) {
      throw new Error('Invalid port range');
    }
    const ports = [];
    for (let p = start; p <= end; p++) {
      ports.push(p);
    }
    return ports;
  } else {
    return portArg.split(',').map(p => {
      const port = Number(p.trim());
      if (isNaN(port)) {
        throw new Error('Invalid port list');
      }
      return port;
    });
  }
}

function scanPort(host, port, timeout = 1000) {
  return new Promise((resolve) => {
    const socket = new net.Socket();
    let status = 'closed';

    socket.setTimeout(timeout);

    socket.on('connect', () => {
      status = 'open';
      socket.destroy();
    });

    socket.on('timeout', () => {
      socket.destroy();
    });

    socket.on('error', () => {
    });

    socket.on('close', () => {
      resolve({ port, status });
    });

    socket.connect(port, host);
  });
}

function outputResults(results, format) {
  if (format === 'json') {
    const json = JSON.stringify(results, null, 2);
    fs.writeFileSync('scan_results.json', json);
    console.log('Results saved to scan_results.json');
  } else if (format === 'csv') {
    const header = 'port,status\n';
    const rows = results.map(r => `${r.port},${r.status}`).join('\n');
    fs.writeFileSync('scan_results.csv', header + rows);
    console.log('Results saved to scan_results.csv');
  } else {
    results.forEach(r => {
      console.log(`Port ${r.port}: ${r.status}`);
    });
  }
}

async function main() {
  const [,, host, portsArg, format] = process.argv;

  if (!host || !portsArg) {
    console.error('Usage: node portscanner.js <host> <port(s)> [json|csv]');
    console.error('Ports can be a range (20-80) or comma separated (22,80,443)');
    process.exit(1);
  }

  let ports;
  try {
    ports = parsePorts(portsArg);
  } catch (err) {
    console.error('Error parsing ports:', err.message);
    process.exit(1);
  }

  console.log(`Scanning ${host} on ports: ${ports.join(', ')}`);

  const results = [];
  for (const port of ports) {
    const res = await scanPort(host, port);
    results.push(res);
  }

  outputResults(results, format);
}

main();

File: portscanner.py
import sys
from scapy.all import IP, TCP, sr1

def parse_ports(port_arg):
    ports = []
    for part in port_arg.split(','):
        if '-' in part:
            start, end = map(int, part.split('-'))
            ports.extend(range(start, end + 1))
        else:
            ports.append(int(part))
    return ports

def syn_scan(host, port, timeout=1):
    packet = IP(dst=host)/TCP(dport=port, flags="S")
    response = sr1(packet, timeout=timeout, verbose=0)
    if response is None:
        return {'port': port, 'status': 'filtered/no response'}
    if response.haslayer(TCP):
        if response[TCP].flags == 0x12:
            rst = IP(dst=host)/TCP(dport=port, flags="R")
            sr1(rst, timeout=timeout, verbose=0)
            return {'port': port, 'status': 'open'}
        elif response[TCP].flags == 0x14:
            return {'port': port, 'status': 'closed'}
    return {'port': port, 'status': 'unknown'}

def main():
    if len(sys.argv) < 3:
        print("\033[1;32mUsage: python portscanner.py <host> <ports>\033[0m")
        print("Ports can be a single number, comma-separated, or ranges (example: 22,80,100-110)")
        sys.exit(1)

    host = sys.argv[1]
    ports_arg = sys.argv[2]
    ports = parse_ports(ports_arg)

    print(f"Performing SYN scan on {host} for ports: {', '.join(map(str, ports))}")
    for port in ports:
        result = syn_scan(host, port)
        print(f"Port {port}: {result['status']}")

if __name__ == "__main__":
    main()
